draft-ietf-ppm-dap.md

```diff
@@ -1391,124 +1317,230 @@ Otherwise, `out` is the pair `(prep_state, prep_msg)`.
+the Leader. The order of the Helper's `prepare_steps` MUST match that of the
+Leader's. The Helper then awaits the next message from the Leader.
 
-* If failed, then mark the report as failed and reply with a failed PrepareStep
-  to the leader.
-* If finished, then mark the report as finished and reply with a finished
-  PrepareStep to the leader. The helper then stores the output share and awaits
-  for collection; see {{collect-flow}}.
+[[OPEN ISSUE: consider relaxing this ordering constraint. See issue#217.]]
 
-Otherwise, preparation continues. In this case, the helper computes its updated
-state and output message as follows:
+### Aggregate Job Resource {#aggregation-job-resource}
+
+This resource allows the Leader to create aggregation jobs in the Helper and
+check on their state. Only the Helper supports this resource.
+
+##### Representation {#aggregation-job-representation}
 
 ~~~
-out = VDAF.prep_next(prep_state, inbound)
-~~~
+opaque AggregationJobId[16];
+
+struct {
+  ReportMetadata metadata;
+  opaque public_share<0..2^32-1>;
+  HpkeCiphertext encrypted_input_share;
+} ReportShare;
+
+struct {
+  QueryType query_type;
+  select (PartialBatchSelector.query_type) {
+    case time_interval: Empty;
+    case fixed_size: BatchID batch_id;
+  };
+} PartialBatchSelector;
 
-where `inbound` is the previous VDAF prepare message sent by the leader and
-`prep_state` is the helper's current preparation state. If this operation fails,
-then the helper fails with error `vdaf_prep_error`. Otherwise, it interprets
-`out` as follows:
+enum {
+  start(0),
+  continued(1),
+  finished(2),
+  failed(3),
+  (255),
+} PrepareStepState;
 
-* If this is the last round of VDAF preparation phase, then `out` is the
-  helper's output share, in which case the helper stores the output share for
-  future collection.
-* Otherwise, the helper interprets `out` as the tuple `(new_state, prep_msg)`,
-  where `new_state` is its updated preparation state and `prep_msg` is its next
-  VDAF message.
+enum {
+  batch_collected(0),
+  report_replayed(1),
+  report_dropped(2),
+  hpke_unknown_config_id(3),
+  hpke_decrypt_error(4),
+  vdaf_prep_error(5),
+  batch_saturated(6),
+  task_expired(7),
+  unrecognized_message(8),
+  report_too_early(9),
+  (255)
+} ReportShareError;
 
-This output message for each report in AggregateContinueReq.prepare_steps is
-then sent to the leader in an AggregateContinueResp message, structured as
-follows:
+struct {
+  ReportID report_id;
+  PrepareStepState prepare_step_state;
+  select (PrepareStep.prepare_step_state) {
+    case start: ReportShare;
+    case continued: opaque prep_msg<0..2^32-1>;
+    case finished: Empty;
+    case failed: ReportShareError;
+  }
+} PrepareStep;
 
-~~~
 struct {
+  u16 round;
+  opaque agg_param<0..2^32-1>;
+  PartialBatchSelector part_batch_selector;
   PrepareStep prepare_steps<1..2^32-1>;
-} AggregateContinueResp;
+} AggregationJob;
```

Comment:
I've been working on implementing this in Janus (https://github.com/divviup/janus/pull/731/files) and I'm finding this definition of `AggregationJob` a little awkward. The idea here is to have a single representation of an aggregation job for both helper and leader across multiple rounds of preparation. But we end up with some extra complexity in implementations:

- Not all `PrepareStep.prepare_step_state` values are valid in all rounds of preparation. For instance, when the leader first sends an `AggregationJob` to the helper, the helper must verify that each `PrepareStep` is in the `start` state. In subsequent rounds, the `start` state is illegal.
- `PrepareStep`s in the `start` state contain a `ReportShare`, which contains `ReportMetadata`, which has `ReportID` in it. That duplicates the `ReportID` field in the outer `PrepareStep` structure, which is wasteful.
- the `agg_param` and `part_batch_selector` fields are only really useful in the first leader->helper messages. Subsequently, the helper has to check that they haven't changed since an earlier round.

All of these mean extra code for servers to write to validate messages. I think we were better off the way things were before, with distinct representations for the initial creation of an aggregation job and then the subsequent stepping. I think it's less REST-y, but that doesn't seem like a good enough tradeoff to me against the implementation complexity and increased message size.

Reply:
+1, I think we should go back to what we had before (more or less). My main concern: We should not transmit the agg param more than once per agg job, since for Poplar this might be relatively large. (Think 1-2MBs.)